pkg/sharky/store.go

```diff
@@ -84,64 +94,32 @@ func (s *Store) create(index uint8, maxDataSize int, basedir fs.FS) (*shard, err
 	if err != nil {
 		return nil, err
 	}
-	sl := newSlots(ffile.(sharkyFile), s.wg)
+	sl := newSlots(ffile.(sharkyFile))
 	err = sl.load()
 	if err != nil {
 		return nil, err
 	}
 	sh := &shard{
-		reads:       make(chan read),
-		errc:        make(chan error),
-		writes:      s.writes,
+		available:   s.availableShards,
 		index:       index,
 		maxDataSize: maxDataSize,
 		file:        file.(sharkyFile),
 		slots:       sl,
 		quit:        s.quit,
 	}
-	terminated := make(chan struct{})
-	sh.slots.wg.Add(1)
+	s.wg.Add(1)
 	go func() {
-		defer sh.slots.wg.Done()
+		defer s.wg.Done()
 		sh.process()
```

Comment:
Currently new goroutine is started for each `shard` instance, would it be possible to avoid these goroutines? Are they really necessary?

--- 

All that `sh.process()` function does is essentially two operations:

```
slot := sh.slots.Next()
sh.slots.Use(slot)
```

Would it be possible to just call these two functions in (store.go:136) instead of reading from channel? Probably one more utility function would be needed to get shard, but overall code might be more simpler.

Reply:
good idea actually :)